### ISSUE TITLE
Multiple selections via shift+select don't work

### ISSUE THREAD

**Issue by canavandl** [bokeh/bokeh#6032]
v bokeh 0.12.5dev16
python3
Chrome 57

Confirmed for the LassoSelectTool and BoxSelectTool, but thought to exist for all selection tools.

Best to reproduce with examples/app/selection_histogram.py

After doing a first box selection (which works normally), holding shift to perform another box selection causes the second selection to not "end" (i.e. the overlay will continue following the cursor). The glyphs in the second selection area will be rendering as their selection_glyphs after ending the second selection, but it seems that some "selected" event isn't occuring, because the selection histogram app demo doesn't update the histograms.

**Comment by bryevdv** [comment]
> holding shift to perform another box selection causes the second selection to not "end" (i.e. the overlay will continue following the cursor)

FWIW I can't reproduce this aspect. But it is failing to respect the multiple selection (just makes an entirely new one)

**Comment by mattpap** [comment]
@bryevdv, this is broken since protocol changes. There are some `undefined` values floating around causing problem. I'm investigating what's the root cause.

**Comment by bryevdv** [comment]
This is well before that. It's broken with `0.12.5dev1`

**Comment by bryevdv** [comment]
This is the commit that broke things:

https://github.com/bokeh/bokeh/commit/3995f0f67c0a2a090d06b7b5e50ca31e654da131

**Comment by bryevdv** [comment]
I should add: it's possible there are two problems. The problem with "not ending" might be a newer protocol problem. But shift being ignored goes back to the commit above, which does *not* have the "not ending" problem.

**Comment by mattpap** [comment]
OK, both changes are at fault. Protocol changes, because there is `v.buffer` code and `v` can be `undefined`, but `null` is only considered (I even asked if it's possible to have `v` undefined). 3995f0f for some reason creates arrays like `[undefined]` for indices.

**Comment by bryevdv** [comment]
I propose that:

* you look into the protocol issue
* finish looking into the indices union issue
* luke makes a selenium test when they are both fixed

**Comment by mattpap** [comment]
@bryevdv, I'm already working on a PR.

**Comment by canavandl** [comment]
great - i'll push a single/few selenium tests to matt's PR when he's done.

**Comment by bryevdv** [comment]
For both issues?

FYI when this is called:
```
indices['2d'].indices = array_1.union(this.indices['2d'].indices, indices['2d'].indices);
``` 
both arguments are undefined. I'd argue that `union` probably ought to return `undefined` in that case tho we could also check the args before unioning

**Comment by bryevdv** [comment]
The structure of 2d indices is not a simple list BTW so I am not sure a union there is appropriate at all.

**Comment by mattpap** [comment]
@bryevdv, not at all, but that was like this for ages.

**Comment by bryevdv** [comment]
Certainly, but whatever underscore union did and your union are different and exposed it. Removing that line solves the "union" problem I'd suggest that as the immediate solution. I would not expect multiple 2d selections to have been or currently be working anyway.

**Comment by bryevdv** [comment]
removing that line solves both issues.

**Comment by mattpap** [comment]
@bryevdv, fix available in PR #6033.
